Destroy MIMEFieldBlockImpl that doesn't have fields in use

maskit opened the pull request:
Because MIMEFieldBlocks are destroyed only when MIMEHdr is destroyed, the block
chain never be short. With this change, a block will be destroyed when all fields
in a block are deleted, and the block will be removed from a block chain.

A chain can be very long on a long live H2 session because HPACK uses MIMEHdr hard.

---

~This is an early preview and WIP. I wanted to gather feedback before making many changes.~

maskit commented:
I guess it was by design because there was no use-case that requires thousands of add/delete operations against a MIMEHdr, and destroying blocks could be (small) overhead. But I think keeping a chain short is worth now because HPACK uses MIMEHdr hard.

SolidWallOfCode commented:
Two things:

*  When the block is deleted, the slotnum of subsequent fields changes. Are these stored anywhere? There is `_mime_hdr_field_list_search_by_slotnum` - not sure what calls that.

*  Should we perhaps just switch to `std::deque` and pop dead elements.

maskit commented:
> When the block is deleted, the slotnum of subsequent fields changes. Are these stored anywhere? There is _mime_hdr_field_list_search_by_slotnum - not sure what calls that.

Good point. Currently a slot number are checked right before use so it should not be a problem now, but there is no guarantee. I'm not sure if we should have it as a kind of constraint.

Also, I found another use of slotnum, and it can brake something because slot number 0-31 are special. `mime_hdr_get_accelerator_slotnum` does some calculation based on a slot number. We probably be able to handle it by not removing the first two blocks.

> Should we perhaps just switch to std::deque and pop dead elements.

Sounds reasonable to me, but it would require a big change and I don't have enough motivation on it.

SolidWallOfCode commented:
I wonder if we could change the # of fields to block to something larger like 32. Then all of the special fields could be in the first block.
I am concerned about looping over the blocks. Is the assumption the block list will always be short enough this isn't an issue?
On a deeper level, why is HPACK deleting fields?

maskit commented:
> I wonder if we could change the # of fields to block to something larger like 32. Then all of the special fields could be in the first block.

I think it's possible but it would need to be done very carefully. I'm not confident about the change.

> I am concerned about looping over the blocks. Is the assumption the block list will always be short enough this isn't an issue?

Yes, that is my assumption. It works for HPACK at minimum because it works like FIFO.

> On a deeper level, why is HPACK deleting fields?

HPACK uses MIMEHdr for its Dynamic Table. Headers in Dynamic Table will be deleted when a new header has been inserted into the table and the total header size exceeds configured limit, which is required by the spec. In short, it keeps only headers recently inserted.

SolidWallOfCode commented:
Looking deeper in the code about this, I don't think the accelerator slotnums are a real issue. The slotnum values for those are limited to the first block, and that is never deleted because it's previous pointer is `nullptr`. Thinking about that, though, you could be a bit more efficient if you checked `_prev == nullptr` first, before checking if all of the fields in the block are deleted. Or alternatively, have the block loop start at `m_first_block->_next`.

maskit commented:
> I don't think the accelerator slotnums are a real issue. The slotnum values for those are limited to the first block

Yeah, so it's "keep the first two blocks" or "increase a number of slots in a block". And the latter sounds nice to me too, but I have no idea what it means. In other words, why is it currently 16?

SolidWallOfCode commented:
I don't see why you'd need to keep the first two blocks. The slotnum in the accelerator slots is limited to the first block (it's only 4 bits). There are 32 accelerator slots, but that's a different thing. Just the first block should be sufficient.

I think it's 16 because reasons. I can't find any fundamental reason for it. It's a tradeoff between wasted space and extra allocations. Back then,  I suspect having more than 16 headers was very unusual. Not so much today. But you're right that we should do that in a separate PR if it looks like a good idea.

maskit commented:
Alright, I should probably look into the accelerator logic again.

I'm going to make changes below:
- Make sure fields will not be accessed with an old (invalid) slotnum
- Check `_prev == nullptr` for efficiency

Use of `std::dequeue` and increasing a number of slots would be separate PRs if we do.

maskit commented:
- Confirmed that slotnum returned from `mime_hdr_field_slotnum` is not kept now, and added a comment to the function so that nobody will keep it in the future
- Added a condition that checks `prev_block` for efficiency -- didn't change the start block to make it easy to read
- Removed redundant code

maskit commented:
@SolidWallOfCode 

> I don't see why you'd need to keep the first two blocks.

You're right. Just one block is enough.

maskit commented:
Please take another look.

maskit commented:
I think I should probably put a bit more of information to have your attention. This is a fix for a performance issue but not a trivial refactoring.

maskit commented:
[approve ci]

maskit commented:
Sigh, clang-analyzer job fails now. It passed when I pushed it...

SolidWallOfCode commented on proxy/hdrs/MIME.cc:
```diff
@@ -1635,20 +1635,8 @@ mime_hdr_field_delete(HdrHeap *heap, MIMEHdrImpl *mh, MIMEField *field, bool del
-      mime_field_destroy(mh, field);
-
-      MIME_HDR_SANITY_CHECK(mh);
+      mime_hdr_field_delete(heap, mh, field, false);
```
Two things
*  This will not track `m_lost_string_space` correctly. I don't know if that's important with the other changes.
*  Will the duplicate pointer on the first (non-duplicate) field be updated correctly? Is that done by the caller?

maskit commented:
[approve ci Clang-Analyzer]

zwoop commented:
Cherry-picked to v9.0.x branch.

zwoop commented:
This has merge conflicts, and we'll need a new 8.1.x PR. Also, feel free to recommend / suggest other PRs that this depends on, such that it makes the back porting easier.
